addons/mrp/models/mrp_bom.py

```diff
@@ -613,14 +613,39 @@ def create(self, vals_list):
                 values['product_uom_id'] = self.env['product.product'].browse(values['product_id']).uom_id.id
         return super(MrpBomLine, self).create(vals_list)
 
-    def _skip_bom_line(self, product):
-        """ Control if a BoM line should be produced, can be inherited to add
-        custom control.
+    def _skip_bom_line(self, product, never_attribute_values):
+        """ Control if a BoM line should be produced, can be inherited to add custom control.
+            cases:
+                - no_variant:
+                    1. attribute present on the line
+                        => need to be at least one attribute value matching between the one passed as args and the ones one the line
+                    2. attribute not present on the line
+                        => valid if the line has no attribute value selected for that attribute
+                - always and dynamic: match_all_variant_values()
         """
         self.ensure_one()
         if product._name == 'product.template':
             return False
-        return not product._match_all_variant_values(self.bom_product_template_attribute_value_ids)
+
+        # attributes create_variant 'always' and 'dynamic'
+        other_attribute_valid = product._match_all_variant_values(self.bom_product_template_attribute_value_ids.filtered(lambda a: a.attribute_id.create_variant != 'no_variant'))
+
+        if not self.bom_product_template_attribute_value_ids.filtered(lambda a: a.attribute_id.create_variant == 'no_variant'):
+            return not other_attribute_valid
+
+        # No_variant attribute checks
+        never_attribute_values = never_attribute_values or self.env['product.template.attribute.value']
+        attribute_ids = (self.bom_product_template_attribute_value_ids + never_attribute_values).filtered(
+                lambda a: a.attribute_id.create_variant == 'no_variant'
+            ).mapped('attribute_id')
+
+        is_attribute_ids_valid = all(
```

Comment:
maybe make it more readable with explicit loop. To try

Reply:
groupby the value by attribute rather than browsing the list to find them each time